C++ add move constructor to MessageBuilder

sutambe opened the pull request:
This PR adds a move ctor, move assign, and swap in `MessageBuilder`. I've a use-case where I don't know the run-time size of `SliceAllocator::slice_`. I'm using a default `MessageBuilder`. I would like to move assign a new `MessageBuilder` when the dynamic size of the `slice_` is known.

For move constructor, default construct and swap idiom is used. For move assignment operator, construct a temporary and swap idiom is used. So as long as `swap` is well-behaved, things should work out. As `grpc_slice` is a value type, `std::swap` should do the right thing.

If the PR looks reasonable, I think of a way to add unit tests. Suggestions welcome.

sutambe commented:
I signed the Google Individual Contributor License Agreement.

aardappel commented:
This certainly looks reasonable.. the amount of verbosity is unfortunate but unavoidable I agree.

I don't think we have any GRPC specific unit tests yet, in fact, we don't even run GRPC tests by default, since we don't want to depend on GRPC being present. But adding some, that could maybe be run as part of the existing grpctest.cpp would be very welcome.

aardappel commented:
I remember looking at hunter once, and it seems like a cool idea. Still, supporting something which is not likely to be used by other GRPC users is not ideal.

aardappel commented:
The main `CMakeLists.txt` has `option(FLATBUFFERS_BUILD_GRPCTEST "Enable the build of grpctest" OFF)`

sutambe commented:
Hi @aardappel, I think I've something now that's ready for review. I've
1. Added move semantics to `MessageBuilder`, `FlatBufferBuilder`, `SliceAllocator`, and `vector_downward`
2. Added unit tests
3. Updated flatbuffers + gRPC build instructions
4. Updated `CMakeLists.txt` with simplified build instructions.

The general pattern is _default/move construct and swap_ idiom. I think it works well for these classes because some of them have up to 6 members and `swap` helps reduce verbosity. More importantly,  no conditional `delete` calls for `string_pool` and `allocator` are needed in `FlatBufferBuilder` and `vector_downward`, respectively. The "right" things happen as a side-effect of `temp` going out of scope.

The unit tests check the same invariants for `MessageBuilder`, `FlatBufferBuilder`, and a test-only message builder that uses heap allocation (`HeapMessageBuilder`). 

@aardappel Who else do you think should review this patch?

aardappel commented:
It would be cool if @llchan could review, since he is responsible for a lot of previous gRPC and allocator improvements.

aardappel commented on CMakeLists.txt:
```diff
@@ -257,6 +258,9 @@ if(FLATBUFFERS_BUILD_GRPCTEST)
   endif()
+  INCLUDE_DIRECTORIES(/your/path/to/grpc/install/include)
+  INCLUDE_DIRECTORIES(/your/path/to/protobuf/include)
+  LINK_DIRECTORIES(/your/path/to/grpc/install/lib)
```
Why do we have these here? If they are meant as documentation they should maybe be comments?
Last I worked on this code, such include and link dirs are not necessary if you `make install` gRPC, if I remember correctly.

sutambe replied:
You may be right. I did not `make install` gRPC in the system directory. C++ gRPC docs warn against that saying uninstalling will not be easy. So I did `make prefix= $(pwd)/install install`. I'm using includes and libs off of that. Hence the paths.

Just compiling and installing grpc was not enough. While compile `grpc_test.cpp` I got a compiler error that `common.h` is missing
```
Scanning dependencies of target grpctest
[ 71%] Building CXX object CMakeFiles/grpctest.dir/tests/monster_test.grpc.fb.cc.o
In file included from /home/sutambe/projects/my-oss/grpc/install/include/grpcpp/impl/codegen/proto_utils.h:28:0,
                 from /home/sutambe/projects/my-oss/grpc/install/include/grpc++/impl/codegen/proto_utils.h:26,
                 from /home/sutambe/projects/my-oss/myflatbuffers/tests/monster_test.grpc.fb.h:13,
                 from /home/sutambe/projects/my-oss/myflatbuffers/tests/monster_test.grpc.fb.cc:6:
/home/sutambe/projects/my-oss/grpc/install/include/grpcpp/impl/codegen/config_protobuf.h:25:42: fatal error: google/protobuf/stubs/common.h: No such file or directory
 #include <google/protobuf/stubs/common.h>
                                          ^
compilation terminated.
make[2]: *** [CMakeFiles/grpctest.dir/tests/monster_test.grpc.fb.cc.o] Error 1
make[1]: *** [CMakeFiles/grpctest.dir/all] Error 2
make: *** [all] Error 2
```

It's kind of weird that protobuf headers are needed while compiling flatbuffer tests. The fix could be in a separate PR though.

The `INCLUDE_DIRECTORIES` and `LINK_DIRECTORIES` serve dual purpose: CMake code and comments. I'm fine with proper cmake variables. I was wondering about them myself.

aardappel commented on grpc/README.md:
```diff
@@ -9,3 +9,26 @@ from GRPC, and work with both the Protobuf and FlatBuffers code generator.
+### Linux
+
+1. Build and install gRPC. See [instructions](https://github.com/grpc/grpc/tree/master/src/cpp).
+2. Build and install protobuf. See [instructions](https://github.com/protocolbuffers/protobuf/blob/master/src/README.md).
```
Does step 2 not happen automatically as a consequence of 1? I never remember building protobuf manually.

sutambe replied:
No, it was not automatic. I had to do two separate build and install. I'll give it a shot on a different Linux distro.

aardappel commented on grpc/README.md:
```diff
@@ -9,3 +9,26 @@ from GRPC, and work with both the Protobuf and FlatBuffers code generator.
+1. Build and install gRPC. See [instructions](https://github.com/grpc/grpc/tree/master/src/cpp).
+2. Build and install protobuf. See [instructions](https://github.com/protocolbuffers/protobuf/blob/master/src/README.md).
+   A gRPC header depends on a Protobuf header file.
+3. In the `if(FLATBUFFERS_BUILD_GRPCTEST)` block in the root `CMakeLists.txt`, update
```
If this is actually needed (which I don't think it is, see above), it would be better to create a CMake variable for the install path, and make all of these relative to it, such that users don't need to modify this file.

aardappel commented on grpc/README.md:
```diff
@@ -9,3 +9,26 @@ from GRPC, and work with both the Protobuf and FlatBuffers code generator.
+
+### Linux
+
+1. `ln -s /your/path/to/grpc/install/lib/libgrpc++_unsecure.so.6 /your/path/to/grpc/install/lib/libgrpc++_unsecure.so.1`
```
none of this was needed last time I tried.. just `./grpctest`.

aardappel commented on grpc/tests/message_builder_test.cpp:
```diff
@@ -0,0 +1,148 @@
+
+struct TestHeapMessageBuilder : public flatbuffers::FlatBufferBuilder {
+  TestHeapMessageBuilder()
+    : flatbuffers::FlatBufferBuilder(2048, new flatbuffers::DefaultAllocator(), true) {}
```
Why are you allocating a `DefaultAllocator` explicitly?

sutambe replied:
There are three cases
1. `FlatBufferBuilder` with a null allocator--`FlatBufferBuilder` defaults to `DefaultAllocator` internally.
2. `MessageBuilder` with a funky allocator that's a member--`FlatBufferBuilder` uses the passed allocator in non-owned manner.
3. `TestHeapMessageBuilder` with a custom allocator (owned) that's neither option 1 nor 2. I used `DefaultAllocator` as it was handy. It could be anything that `is-a` `flatbuffers::Allocator`.

May be this is clearer.
```cpp
struct OwnedAllocator : public flatbuffers::DefaultAllocator {};

struct TestHeapMessageBuilder : public flatbuffers::FlatBufferBuilder {
  TestHeapMessageBuilder()
    : flatbuffers::FlatBufferBuilder(2048, new OwnedAllocator(), true) {}
};

aardappel commented on grpc/tests/message_builder_test.cpp:
```diff
@@ -0,0 +1,148 @@
+  }
+};
+
+void builder_tests() {
```
Nice!

sutambe commented:
@llchan Do you have some cycles to look at this patch? If not, @aardappel do you think it's good to go in? I'll squash the PR to a single commit.

aardappel commented on CMakeLists.txt:
```diff
@@ -257,6 +258,9 @@ if(FLATBUFFERS_BUILD_GRPCTEST)
   if(CMAKE_COMPILER_IS_GNUCXX)
     set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-unused-parameter -Wno-shadow")
   endif()
+  INCLUDE_DIRECTORIES(${GRPC_INSTALL_PATH}/include)
```
If these env vars are not specified, we get.. `/include` ?

sutambe replied:
Yes. I'll add `if` to bail out if the vars are not defined.

aardappel commented on grpc/README.md:
```diff
@@ -9,3 +9,23 @@ from GRPC, and work with both the Protobuf and FlatBuffers code generator.
+
+1. Download, build and install gRPC. See [instructions](https://github.com/grpc/grpc/tree/master/src/cpp).
+    * Lets say your gRPC clone is at `/your/path/to/grpc_repo`.
+    * Install gRPC in a custom directory by running `make install prefix=/your/path/to/grpc_repo/install`.
```
Like I mentioned earlier, if you a regular `make install` without the custom dir, none of these environment vars appear to be necessary, is that the case? Should that be mentioned in this doc as an easier way? Also, I don't remember having to build protobuf seperately.

sutambe replied:
It's not possible to get that far if the includes are missing. Compilations fails in `make` with the following error on Ubuntu 18. Same story on RHEL7. 

Building protobuf is not necessary. I've updated the instructions to reflect that. Include path, however is necessary. 
```
$ cmake -DFLATBUFFERS_BUILD_GRPCTEST=ON ..
$ make
...
[ 45%] Building CXX object CMakeFiles/grpctest.dir/tests/monster_test.grpc.fb.cc.o
In file included from /home/sutambe/projects/myoss/myflatbuffers/tests/monster_test.grpc.fb.h:8:0,
                 from /home/sutambe/projects/myoss/myflatbuffers/tests/monster_test.grpc.fb.cc:6:
/home/sutambe/projects/myoss/myflatbuffers/include/flatbuffers/grpc.h:23:10: fatal error: grpc++/support/byte_buffer.h: No such file or directory
 #include "grpc++/support/byte_buffer.h"
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
CMakeFiles/grpctest.dir/build.make:66: recipe for target 'CMakeFiles/grpctest.dir/tests/monster_test.grpc.fb.cc.o' failed
make[2]: *** [CMakeFiles/grpctest.dir/tests/monster_test.grpc.fb.cc.o] Error 1
CMakeFiles/Makefile2:67: recipe for target 'CMakeFiles/grpctest.dir/all' failed
make[1]: *** [CMakeFiles/grpctest.dir/all] Error 2
Makefile:140: recipe for target 'all' failed
make: *** [all] Error 2
```

aardappel replied:
fair enough.. we can merge then.

sutambe commented:
Checkout [my personal travis build](https://travis-ci.com/sutambe/flatbuffers/builds/83428452) testing the changes in the `CMakeLists.txt`. Otherwise, this build is completely unrelated to this PR.

aardappel commented:
Thanks!
